test/utils/test_common.py

```diff
@@ -70,3 +77,108 @@ def test_int(self, expected):
     def test_tuple(self, expected):
         actual = to_tuple_tuple((2, 2, 2), 3, 3)
         assert actual == expected, "tuple -> tuple[tuple] incorrect"
+
+
+class TestCheckpointWrapper:
+    @pytest.fixture
+    def model(self):
+        class DummyAttention(nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.param = nn.Parameter(torch.ones(3))
+
+            def forward(self, x, y, attn_mask=None, use_cache=None):
+                grad = x if use_cache else y
+                grad = grad * self.param
+                if attn_mask is not None:
+                    grad = grad * attn_mask
+                return grad
+
+        class DummyIdentity(nn.Module):
```

Comment:
Just wondering, why do we use DummyIdentity instead of nn.Identity here? (I know you had nn.Identity previously, so assuming there is some reason for the change)

Reply:
Good Q. Because `nn.Identity` does not render the output to be `requires_grad`. In order for the checkpointed attention layer to receive a "gradient-ed" input, I have to manually craft the identity layer.